### PR TITLE
Re-work pack lowering

### PR DESCRIPTION
**Pull request by chelini**
Rewriting `tensor.pack` to loops was not composable, preventing further optimizations like fusion. A better approach is to materialize loops only during tiling and fusion. The pack is then generalized to an expanded shape and a transpose. Further optimizations are added to map the transpose and the expand shape to xsmm vnni calls.

### CODE REVIEW

**Review by adam-smnk — COMMENTED** [review]
Should there be any improvement in any benchmarks?

File: lib/TPP/LowerPacksAndUnpacks.cpp [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    if (tileSizes.empty())
+      continue;
+    auto tilingResult = tileOp(
+        rewriter, cast<TilingInterface>(packOp.getOperation()), tileSizes);
+    if (failed(tilingResult))
+      continue;
+    rewriter.replaceOp(packOp, tilingResult->replacements);
+  }
+
+  // Tile unpacks.
```
Comment (adam-smnk) [inline lib/TPP/LowerPacksAndUnpacks.cpp:130]: Is unpack fusion coming in this or separate PR?

**Review by adam-smnk — COMMENTED** [review]
Looks good after the first pass

File: benchmarks/config/base/base.json [hunk 2]
```diff
@@ -46,6 +46,13 @@
       "flags": [ "-n", "100" ],
       "extensions": [ "(avx2|asimd)" ]
     },
+    "fp32_3x512_mlir": {
```
Comment (rengolin) [inline benchmarks/config/base/base.json:49]: What do these add over the existing 1024 case?

File: benchmarks/config/base/base.json [hunk 3]
```diff
@@ -60,6 +67,13 @@
       "flags": [ "-n", "100" ],
       "extensions": [ "avx512.*" ]
     },
+    "f32_flat_3x1024_unp_mlir": {
```
Comment (rengolin) [inline benchmarks/config/base/base.json:70]: We really need to be moving off hand-crafted MLIR files and use `mlir-gen` for everything.

File: lib/TPP/LowerPacksAndUnpacks.cpp [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+  }
+  return tilingResult;
+}
+
+// Fuse producer and consumer pack. Standalone packs are tiled into 2d tiles.
+static void fuseOrTilePacks(RewriterBase &rewriter, FunctionOpInterface func) {
+  SmallVector<tensor::PackOp> consumerPackOps;
+  SmallVector<tensor::PackOp> otherPacks;
+  SmallVector<tensor::UnPackOp> unPacks;
+  func->walk<WalkOrder::PostOrder>([&](tensor::PackOp consumerPackOp) {
```
Comment (rengolin) [inline lib/TPP/LowerPacksAndUnpacks.cpp:77]: Nit, it took me a while to figure that `consumer` here means "of another pack", not just any other op. I'd call it `chained` instead, to make it clearer.

File: lib/TPP/LowerPacksAndUnpacks.cpp [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    if (producerPackOp)
+      consumerPackOps.push_back(consumerPackOp);
+    else
+      otherPacks.push_back(consumerPackOp);
+  });
+  func->walk<WalkOrder::PostOrder>(
+      [&](tensor::UnPackOp unpackOp) { unPacks.push_back(unpackOp); });
+
+  // Tile and fuse.
+  for (auto consumerPackOp : consumerPackOps) {
```
Comment (rengolin) [inline lib/TPP/LowerPacksAndUnpacks.cpp:90]: Are you assuming here that the chained packs (consumers) are handled before the non-chained (producer) ones?

If you have a 3-chain, wouldn't the second be in the list before the third, breaking that rule?

File: lib/TPP/LowerPacksAndUnpacks.cpp [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+            cast<tensor::ExtractSliceOp>(tiledPack.getSource().getDefiningOp()),
+            forLoops);
+    if (!fusedProducer)
+      continue;
+    rewriter.replaceOp(consumerPackOp, tilingResult->replacements);
+  }
+
+  // Tile packs.
+  for (auto packOp : otherPacks) {
+    SmallVector<int64_t> tileSizes = getTileSizes(packOp);
```
Comment (rengolin) [inline lib/TPP/LowerPacksAndUnpacks.cpp:120]: Shouldn't you be using the same tiles as the consumer? There doesn't seem to be any guarantee to this here or in `getTileSizes`.

Comment (chelini) [reply lib/TPP/LowerPacksAndUnpacks.cpp:120]: Here I don't have any consumer. All the packs are standalone.

File: test/Passes/pass-pack-lowering.mlir
Comment (adam-smnk) [inline test/Passes/pass-pack-lowering.mlir:1]: It'd be nice to have more fusion tests also for more complex chains.

File: benchmarks/config/base/base.json (hunk 2 above)
Comment (chelini) [reply benchmarks/config/base/base.json]: removed.

File: benchmarks/config/base/base.json (hunk 3 above)
Comment (rengolin) [reply benchmarks/config/base/base.json:63]: No, just move it up and keep a consistent naming. I'll refactor this in one single PR later.

Comment (chelini) [reply benchmarks/config/base/base.json]: moved up, the name is consistent with the other unpack we have, no?

Comment (rengolin) [reply benchmarks/config/base/base.json]: one nit `fp32`, otherwise ok.

**Review by adam-smnk — APPROVED** [review]

**Pull request merged**